components/submit-expense/useExpenseForm.ts

```diff
@@ -0,0 +1,399 @@
+          accountingCategoryId: z.string(),
+        });
+      }
+    }
+
+    if ((query.data?.account?.supportedExpenseTypes ?? []).length > 0) {
+      options.schema = options.schema.extend({
+        expenseTypeOption: z
+          .nativeEnum(ExpenseTypeOption)
+          .refine(v => query.data.account.supportedExpenseTypes.includes(expenseTypeFromOption(v)), {
+            message: 'expense type not supported',
+          }),
+      });
+      options.supportedExpenseTypes = query.data.account.supportedExpenseTypes;
+    }
+
+    if (query.data?.loggedInAccount) {
+      options.payoutProfiles = getPayoutProfiles(query.data.loggedInAccount);
+    }
+
+    options.supportedCurrencies = Object.values(Currency);
+    if (query.data?.payee && values.payoutMethodId && values.expenseTypeOption) {
+      const supportedCurrencies = getSupportedCurrencies(query.data?.account, {
+        payee: query.data?.payee,
+        payoutMethod: query.data?.payee?.payoutMethods?.find(p => p.id === values.payoutMethodId),
+        type: expenseTypeFromOption(values.expenseTypeOption),
+        currency: account.currency,
+      });
+
+      options.schema = options.schema.extend({
+        expenseCurrency: z.string().refine(v => supportedCurrencies.includes(v), {
+          message: `Currency must be one of: ${supportedCurrencies.join(',')}`,
+        }),
+      });
+
+      options.supportedCurrencies = supportedCurrencies;
+    }
+
+    options.schema = options.schema.extend({
+      expenseItems: z.array(
+        z.object({
+          description: z.string().min(1),
+          url:
+            expenseTypeFromOption(values.expenseTypeOption) === ExpenseType.RECEIPT
+              ? z.string().url()
+              : z.string().url().optional(),
+          date: z.string(),
+          amount: z.object({
+            valueInCents: z.number(),
+            currency: z.string().refine(v => Object.values(Currency).includes(v as Currency), {
+              message: `Currency must be one of: ${Object.values(Currency).join(',')}`,
+            }),
+          }),
+        }),
+      ),
+    });
+
+    options.allowExpenseItemAttachment = options.allowExpenseItemCurrencyChange =
+      expenseTypeFromOption(values.expenseTypeOption) === ExpenseType.RECEIPT;
+
+    return options;
+  } catch (err) {
+    return options;
+  }
+}
+
+export function useExpenseForm(opts: {
+  initialValues: Parameters<typeof useFormik<ExpenseFormValues>>['0']['initialValues'];
+  onSubmit: Parameters<typeof useFormik<ExpenseFormValues>>['0']['onSubmit'];
+}): ExpenseForm {
+  const apolloClient = useApolloClient();
+  const { LoggedInUser } = useLoggedInUser();
+  const [formOptions, setFormOptions] = React.useState<ExpenseFormOptions>({ schema: minimumSchema });
+
+  const expenseForm: ExpenseFormik = useFormik<ExpenseFormValues>({
+    initialValues: opts.initialValues,
+    async validate(values) {
+      const result = formOptions.schema.safeParse(values);
```

Comment:
@Betree how does this fit with your idea for form validation using zod?
I have made schema dynamic here, based on picked form values to support conditionally setting some fields are required or other props.

For example, `accountingCategoryId` is made required only if the host/collective/etc selections make it so. `expenseCurrency` is limited to the set of supported currencies, etc..

Reply:
It should work out fine, now that https://github.com/opencollective/opencollective-frontend/pull/10098 is merged you should be able to do directly:

```ts
const expenseForm = useFormikZod<ExpenseFormValues>({ initialValues, schema: formOptions.schema });
```